src/app/search/autocomplete.ts

```diff
@@ -253,44 +250,72 @@ const closingQuoteRegex = /\w["']$/;
  */
 function findLastFilter(
   query: string,
-  caretIndex: number
+  caretIndex: number,
+  searchConfig: SearchConfig
 ): {
   term: string;
   index: number;
 } | null {
   // TODO: maybe include non-whitespace after the caret?
   const queryUpToCaret = query.slice(0, caretIndex);
-  const ast = parseQuery(queryUpToCaret);
-
-  // Find all the "bare" keywords at the end of the query. For example if the query is:
-  // name:foo bar baz
-  // then the trailingKeywordStrings are 'bar baz'
-  const trailingKeywordArgs: string[] = [];
-  traverseAST(
-    ast,
-    (filter) => {
-      if (filter.type === 'keyword' && !/\s+/.test(filter.args)) {
-        trailingKeywordArgs.push(filter.args);
-      } else {
-        return false;
+
+  // Find the index where an incomplete filter starts. For example if the query is:
+  // name:"foo" bar baz
+  // then the open keywords are "bar baz"
+  let earliestIncompleteFilter = -1;
+  try {
+    // We can use the query lexer for this to scan through tokens in the query without parsing the whole AST.
+    for (const token of lexer(queryUpToCaret)) {
+      switch (token.type) {
+        case 'filter': {
+          if (
+            // Ignore complete quoted tokens, they're definitively finished.
+            !token.quoted &&
+            // Match either bare words ...
+            (token.keyword === 'keyword' ||
+              // ... or name:foo style keywords
+              (searchConfig.kvFilters[token.keyword]?.format === 'freeform' &&
+                // Unless they already perfectly match a suggestion without quotes, e.g. name:heritage
+                !searchConfig.suggestions.includes(`${token.keyword}:${token.args}`)))
+          ) {
+            // Only set the index for the first keyword (we're looking for the earliest)
+            if (earliestIncompleteFilter < 0) {
+              // Set it to the beginning of the whole token (including the keyword e.g. name:foo)
+              earliestIncompleteFilter = token.startIndex;
+            }
```

Comment:
I was thinking that this could return multiple guesses as to what the last intended filter could be -- right now this regresses something like `adept pali` (previously autocompletes to `adept name:"the palindrome"`, now tries to complete "adept pali" and suggests nothing). And then prefer suggestions that come from longer filters. This could also give suggestions in above cases where the result is `No fake autocomplete for ...`

Reply:
That's a good idea!